server/portal/apps/auth/models.py

```diff
@@ -92,6 +91,46 @@ def client(self):
 
         return client
 
+    def optimized_client(self):
+        """Tapis client to limit one request to Tapis per User.
+
+        :return: Tapis client using refresh token.
+        :rtype: :class:Tapis
+        """
+        client = self.build_client()
+        if self.expired:
+            logger.info('Tapis OAuth token expired')
+            with transaction.atomic():
+                # Get a lock on this user's token row in db.
+                refreshed_token = TapisOAuthToken.objects.select_for_update().filter(user=self.user).first()
+                if self.is_token_expired(refreshed_token.created, refreshed_token.expires_in):
+                    try:
+                        logger.info('Refreshing tapis oauth token')
+                        client.refresh_tokens()
+                    except Exception:
+                        logger.exception('Tapis Token refresh failed')
+                        raise
```

Comment:
On failure, perhaps we should call `logout`

Reply:
The current logic is all in a model, can't do http redirect or control view responses from here. Have to do from view. I setup an custom exception and handled it in Base View to send 401 back to client. On testing, by forcing an error - it turned the 401 to redirect to tapis oauth, but that failed due to CORS policy issue, I have to check if this is local setup or something else.

Reply:
I came to that realization as well today. I tried another solution with DesignSafe, which is to put the refresh logic in a middleware. In CEP we originally moved that logic from middleware to the `client()` method, but I think the solution you propose here might solve the original issue there?

Haven't tested yet, what are your thoughts?
https://github.com/DesignSafe-CI/portal/blob/task/DES-2702--tapis-v3-oauth/designsafe/apps/auth/middleware.py

Reply:
@rstijerina  - sorry for delay in response, I missed this note. 
I looked at the code in that branch. It looks good, one comment on overall integration:
* Should use do this also for logout? 
logout(request)
 return HttpResponseRedirect(reverse('designsafe_auth:login'))
 
Some testing aspects:
* Behavior on xhr requests when tapis token expiry fails. If xhr does not handle 302 cleanly, some extra check and redirect might be needed.
* Walking through the code, it is protected from infinite loop through this(which is good). If refresh fails, goes to login, which has to authenticate with tapis, if authentication works, this middleware immediately returns (because there is no expiry) and move away from this middleware.

Reply:
> Should use do this also for logout?
logout(request)
return HttpResponseRedirect(reverse('designsafe_auth:login'))

Yes, thanks. Added:
https://github.com/DesignSafe-CI/portal/blob/task/DES-2709--v3-apps-views/designsafe/apps/auth/middleware.py

> Behavior on xhr requests when tapis token expiry fails. If xhr does not handle 302 cleanly, some extra check and redirect might be needed.

Can you expand more on this part? Where would tapis token expiry fail?

Reply:
> > Behavior on xhr requests when tapis token expiry fails. If xhr does not handle 302 cleanly, some extra check and redirect might be needed.
> 
> Can you expand more on this part? Where would tapis token expiry fail?
I meant - "Behavior on xhr requests when tapis token expires, and the refresh fails - this will hit the logout code and send a 302 back to client. If javascript side of response handling does not handle 302 cleanly (page rendering after 302, etc), *may be* extra logic need be needed to check for 302 status and specific error type(token expired) and then setting location href to logout".

Reply:
@rstijerina  - regarding this PR, if middleware is the right place for auth and if it is working in designsafe, should I do the same here and start testing. What is your opinion?

Reply:
The solution I have in DesginSafe does work, here are example logs from a refresh that just occurred for me:
```
des_django         | [DJANGO] INFO 2024-04-12 14:28:57,764 middleware designsafe.apps.auth.middleware.process_request:49: Tapis OAuth token expired for user sal. Refreshing token
des_django         | [DJANGO] INFO 2024-04-12 14:28:57,769 middleware designsafe.apps.auth.middleware.process_request:49: Tapis OAuth token expired for user sal. Refreshing token
des_django         | [DJANGO] INFO 2024-04-12 14:28:57,775 middleware designsafe.apps.auth.middleware.process_request:61: Refreshing Tapis OAuth token
des_django         | [DJANGO] INFO 2024-04-12 14:29:02,626 middleware designsafe.apps.auth.middleware.process_request:72: Token updated by another request. Refreshing token from DB.
```

It might not be fool-proof though, and could definitely use more testing

Reply:
Perhaps we could talk about best place for token refresh in the next infra scrum?